go/vt/vtgate/planbuilder/operators/update.go

```diff
@@ -123,10 +127,54 @@ func createOperatorFromUpdate(ctx *plancontext.PlanningContext, updStmt *sqlpars
 	return buildFkOperator(ctx, op, updClone, parentFks, childFks, vTbl)
 }
 
-func updateWithInputPlanningForFk(ctx *plancontext.PlanningContext, upd *sqlparser.Update) Operator {
+func updateWithInputPlanningRequired(
+	ctx *plancontext.PlanningContext,
+	childFks []vindexes.ChildFKInfo,
+	parentFks []vindexes.ParentFKInfo,
+	updateStmt *sqlparser.Update,
+) bool {
+	if isMultiTargetUpdate(ctx, childFks, parentFks, updateStmt) {
+		return true
+	}
+	// If there are no foreign keys, we don't need to use delete with input.
+	if len(childFks) == 0 && len(parentFks) == 0 {
+		return false
+	}
+	// Limit requires dml with input.
+	if updateStmt.Limit != nil {
+		return true
+	}
+	return false
+}
+
+func isMultiTargetUpdate(ctx *plancontext.PlanningContext, childFks []vindexes.ChildFKInfo, parentFks []vindexes.ParentFKInfo, updateStmt *sqlparser.Update) bool {
+	var targetTS semantics.TableSet
+	for _, ue := range updateStmt.Exprs {
+		targetTS = targetTS.Merge(ctx.SemTable.DirectDeps(ue.Name))
+	}
+	if targetTS.NumberOfTables() == 1 {
+		return false
+	}
```

Comment:
```suggestion
	if ctx.SemTable.Targets.NumberOfTables() == 1 {
		return false
	}
```

Reply:
This cannot be used, as the ctx remains same for all the update queries generated afterwards, so this condition will either remain true or false for all the newer update queries.